### PR TITLE
UPSTREAM: 23145: use versioned object as schema when computing patch on conflict retry

### LINKED ISSUE #6435: Flake in cmd test
With an error like this:

    Running hack/../test/cmd/builds.sh:81: executing 'oc patch bc/ruby-sample-build -p '{"spec":{"output":{"to":{"name":"origin-ruby-sample:latest"}}}}'' expecting success...
    FAILURE after 0.285s: hack/../test/cmd/builds.sh:81: executing 'oc patch bc/ruby-sample-build -p '{"spec":{"output":{"to":{"name":"origin-ruby-sample:latest"}}}}'' expecting success: the command returned the wrong error code
    There was no output from the command.
    Standard error from the command:
    Error from server: unable to find api field in struct BuildConfig for the json field "metadata"
    [FAIL] !!!!! Test Failed !!!!

Occured in PRs #6335 ([build](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8027/consoleFull)), #6409 ([build](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/7986/consoleFull)).

### PR DESCRIPTION
**Pull request by liggitt**
fixes #6435

### CODE REVIEW

**Comment by liggitt** [comment]
this only worked upstream because their internal types still have json tags and still (mostly?) match field structures between internal and external types

need to add a test for origin types

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go [hunk 1]
```diff
@@ -567,11 +567,11 @@ func patchResource(ctx api.Context, admit updateAdmissionFunc, timeout time.Dura
 				return nil, err
 			}
 
-			currentPatch, err := strategicpatch.CreateStrategicMergePatch(originalObjJS, currentObjectJS, patcher.New())
```
Comment (smarterclayton) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go]: I assume this needs to be in Kube 1.2.1 as well.

Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go]: yeah, will open upstream and tag

**Comment by liggitt** [comment]
[test]

**Comment by liggitt** [comment]
@ncdc PTAL

**Comment by liggitt** [comment]
cc @deads2k

**Comment by liggitt** [comment]
Heh, I think my test exceeded the max patch retry limit

**Comment by liggitt** [comment]
@deads2k upstream has LGTM, tag if you're happy with the origin test

**Comment by deads2k** [comment]
[merge]

**Pull request merged**